From 870aa43fa7792dd6f6fc6128f274aea638c6189a Mon Sep 17 00:00:00 2001
From: Wenzong Fan <wenzong.fan@windriver.com>
Date: Mon, 20 Feb 2017 13:42:34 +0000
Subject: [PATCH] refpolicy: some fixes for avc denials

Reference sources: selinux-policy-3.13.1-225.6.fc25.src.rpm
Changes come from: policy-f25-base.patch
                   policy-f25-contrib.patch

* allow radiusd search tmpfs dirs
* allow NetworkManager (gmain) search NetworkManager_initrc_exec dirs
* allow NetworkManager create rawip_socket with type NetworkManager_t
* allow login search tmpfs dirs
* allow ipsec_supervisor (starter) search /run/systemd/journal dir

WRL Fixes:
* allow watchdog search tmpfs dirs
* allow systemd-udevd has capability wake_alarm to itself
* allow systemd-cgroups send messages to unix_dgram_socket
* allow irqbalance has capability dac_override to itself

Fix avc denials:

  avc: denied { search } for pid=334 comm="watchdog" \
  name="/" dev="tmpfs" ino=9358 \
  scontext=system_u:system_r:watchdog_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:tmpfs_t:s0 tclass=dir

  avc: denied { search } for pid=494 comm="radiusd" \
  name="/" dev="tmpfs" ino=9260 \
  scontext=system_u:system_r:radiusd_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:tmpfs_t:s0 tclass=dir

  avc: denied { wake_alarm } for pid=219 \
  comm="systemd-udevd" capability=35 \
  scontext=system_u:system_r:udev_t:s0-s15:c0.c1023 \
  tcontext=system_u:system_r:udev_t:s0-s15:c0.c1023 \
  tclass=capability2

  avc: denied { search } for pid=484 comm="gmain" \
  name="dispatcher.d" dev="sda1" ino=1089942 \
  scontext=system_u:system_r:NetworkManager_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:NetworkManager_initrc_exec_t:s0 tclass=dir

  avc: denied { create } for pid=472 comm="NetworkManager" \
  scontext=system_u:system_r:NetworkManager_t:s0-s15:c0.c1023 \
  tcontext=system_u:system_r:NetworkManager_t:s0-s15:c0.c1023 \
  tclass=rawip_socket

  avc: denied { search } for pid=832 comm="login" \
  name="/" dev="tmpfs" ino=555 \
  scontext=system_u:system_r:local_login_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:tmpfs_t:s0 tclass=dir

  avc: denied { sendto } for pid=1073 comm="systemd-cgroups" \
  path="/run/systemd/cgroups-agent" \
  scontext=system_u:system_r:systemd_cgroups_t:s15:c0.c1023 \
  tcontext=system_u:system_r:init_t:s0-s15:c0.c1023 \
  tclass=unix_dgram_socket

  avc: denied { search } for pid=601 comm="starter" \
  name="journal" dev="tmpfs" ino=12298 \
  scontext=system_u:system_r:ipsec_supervisor_t:s0-s15:c0.c1023 \
  tcontext=system_u:object_r:syslogd_var_run_t:s15:c0.c1023 tclass=dir

  avc: denied { dac_override } for pid=419 comm="irqbalance" \
  scontext=system_u:system_r:irqbalance_t:s0-s15:c0.c1023 \
  tcontext=system_u:system_r:irqbalance_t:s0-s15:c0.c1023 \
  tclass=capability

Upstream-Status: Pending

Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/kernel/files.if            | 18 ++++++++++++++++++
 policy/modules/services/irqbalance.te     |  2 +-
 policy/modules/services/networkmanager.te |  6 +++++-
 policy/modules/services/radius.te         |  1 +
 policy/modules/services/watchdog.te       |  1 +
 policy/modules/system/ipsec.te            |  2 ++
 policy/modules/system/locallogin.te       |  1 +
 7 files changed, 29 insertions(+), 2 deletions(-)

diff --git a/policy/modules/kernel/files.if b/policy/modules/kernel/files.if
index da6fd76..0868024 100644
--- a/policy/modules/kernel/files.if
+++ b/policy/modules/kernel/files.if
@@ -7096,6 +7096,24 @@ interface(`files_dontaudit_search_non_security_dirs',`
 	dontaudit $1 non_security_file_type:dir search_dir_perms;
 ')
 
+########################################
+## <summary>
+##      Allow search the all /var/run directory.
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain to not audit.
+##      </summary>
+## </param>
+#
+interface(`files_search_all_pids',`
+	gen_require(`
+		attribute pidfile;
+	')
+
+	allow $1 pidfile:dir search_dir_perms;
+')
+
 #######################################
 ## <summary>
 ##      Create generic pid directory.
diff --git a/policy/modules/services/irqbalance.te b/policy/modules/services/irqbalance.te
index 6217e0f..ffc5997 100644
--- a/policy/modules/services/irqbalance.te
+++ b/policy/modules/services/irqbalance.te
@@ -24,7 +24,7 @@ init_unit_file(irqbalance_unit_t)
 # Local policy
 #
 
-allow irqbalance_t self:capability { setpcap };
+allow irqbalance_t self:capability { setpcap dac_override };
 dontaudit irqbalance_t self:capability sys_tty_config;
 allow irqbalance_t self:process { getcap getsched setcap signal_perms };
 allow irqbalance_t self:udp_socket create_socket_perms;
diff --git a/policy/modules/services/networkmanager.te b/policy/modules/services/networkmanager.te
index 75ea5cb..2193f36 100644
--- a/policy/modules/services/networkmanager.te
+++ b/policy/modules/services/networkmanager.te
@@ -58,7 +58,7 @@ allow NetworkManager_t self:tun_socket { create_socket_perms relabelfrom relabel
 allow NetworkManager_t self:packet_socket create_socket_perms;
 allow NetworkManager_t self:socket create_socket_perms;
 # ICMPv6 router solicitation/advertisement
-allow NetworkManager_t self:rawip_socket { create setopt getattr write read };
+allow NetworkManager_t self:rawip_socket create_socket_perms;
 
 allow NetworkManager_t wpa_cli_t:unix_dgram_socket sendto;
 
@@ -66,6 +66,10 @@ allow NetworkManager_t NetworkManager_etc_t:dir list_dir_perms;
 allow NetworkManager_t NetworkManager_etc_t:file read_file_perms;
 allow NetworkManager_t NetworkManager_etc_t:lnk_file read_lnk_file_perms;
 
+list_dirs_pattern(NetworkManager_t, NetworkManager_initrc_exec_t, NetworkManager_initrc_exec_t)
+read_files_pattern(NetworkManager_t, NetworkManager_initrc_exec_t, NetworkManager_initrc_exec_t)
+read_lnk_files_pattern(NetworkManager_t, NetworkManager_initrc_exec_t, NetworkManager_initrc_exec_t)
+
 manage_dirs_pattern(NetworkManager_t, NetworkManager_etc_rw_t, NetworkManager_etc_rw_t)
 manage_files_pattern(NetworkManager_t, NetworkManager_etc_rw_t, NetworkManager_etc_rw_t)
 filetrans_pattern(NetworkManager_t, NetworkManager_etc_t, NetworkManager_etc_rw_t, { dir file })
diff --git a/policy/modules/services/radius.te b/policy/modules/services/radius.te
index 1488b4f..7823b6d 100644
--- a/policy/modules/services/radius.te
+++ b/policy/modules/services/radius.te
@@ -96,6 +96,7 @@ domain_use_interactive_fds(radiusd_t)
 
 fs_getattr_all_fs(radiusd_t)
 fs_search_auto_mountpoints(radiusd_t)
+fs_search_tmpfs(radiusd_t)
 
 files_read_usr_files(radiusd_t)
 files_read_etc_runtime_files(radiusd_t)
diff --git a/policy/modules/services/watchdog.te b/policy/modules/services/watchdog.te
index e16ea96..35f245e 100644
--- a/policy/modules/services/watchdog.te
+++ b/policy/modules/services/watchdog.te
@@ -74,6 +74,7 @@ files_read_all_pids(watchdog_t)
 fs_unmount_xattr_fs(watchdog_t)
 fs_getattr_all_fs(watchdog_t)
 fs_search_auto_mountpoints(watchdog_t)
+fs_search_tmpfs(watchdog_t)
 
 auth_append_login_records(watchdog_t)
 
diff --git a/policy/modules/system/ipsec.te b/policy/modules/system/ipsec.te
index 732fed5..caf6f08 100644
--- a/policy/modules/system/ipsec.te
+++ b/policy/modules/system/ipsec.te
@@ -495,6 +495,8 @@ allow ipsec_supervisor_t ipsec_var_run_t:sock_file { rw_sock_file_perms unlink }
 manage_dirs_pattern(ipsec_supervisor_t, ipsec_var_run_t, ipsec_var_run_t)
 manage_files_pattern(ipsec_supervisor_t, ipsec_var_run_t, ipsec_var_run_t)
 files_pid_filetrans(ipsec_supervisor_t, ipsec_var_run_t, { dir file sock_file })
+files_search_all_pids(ipsec_supervisor_t)
+mls_file_read_all_levels(ipsec_supervisor_t)
 
 domtrans_pattern(ipsec_supervisor_t, ipsec_exec_t, ipsec_t)
 
diff --git a/policy/modules/system/locallogin.te b/policy/modules/system/locallogin.te
index 86b034b..0c5b7b8 100644
--- a/policy/modules/system/locallogin.te
+++ b/policy/modules/system/locallogin.te
@@ -104,6 +104,7 @@ files_read_world_readable_sockets(local_login_t)
 files_read_var_symlinks(local_login_t)
 
 fs_search_auto_mountpoints(local_login_t)
+fs_search_tmpfs(local_login_t)
 
 storage_dontaudit_getattr_fixed_disk_dev(local_login_t)
 storage_dontaudit_setattr_fixed_disk_dev(local_login_t)
-- 
2.7.4

